File: collector/analyzer/cpuanalyzer/cpu_analyzer.go
package cpuanalyzer

import (
	"context"
	"fmt"
	"log"
	"os"
	"strconv"
	"sync"

	"github.com/Kindling-project/kindling/collector/analyzer"
	"github.com/Kindling-project/kindling/collector/component"
	"github.com/Kindling-project/kindling/collector/consumer"
	"github.com/Kindling-project/kindling/collector/model"
	"github.com/Kindling-project/kindling/collector/model/constnames"
	"github.com/olivere/elastic/v6"
	"go.uber.org/zap"
)

const (
	CpuProfile analyzer.Type = "cpuanalyzer"
)

type CpuAnalyzer struct {
	cfg          *Config
	cpuPidEvents map[uint32]map[uint32]TimeSegments
	// { pid: routine }
	sendEventsRoutineMap sync.Map
	lock                 sync.Mutex
	esClient             *elastic.Client
	telemetry            *component.TelemetryTools
}

func (ca *CpuAnalyzer) Type() analyzer.Type {
	return CpuProfile
}

func (ca *CpuAnalyzer) ConsumableEvents() []string {
	return []string{constnames.CpuEvent, constnames.JavaFutexInfo, constnames.TransactionIdEvent}
}

func NewCpuAnalyzer(cfg interface{}, telemetry *component.TelemetryTools, consumers []consumer.Consumer) analyzer.Analyzer {
	config, _ := cfg.(*Config)
	ca := &CpuAnalyzer{
		cfg:       config,
		telemetry: telemetry,
	}
	ca.cpuPidEvents = make(map[uint32]map[uint32]TimeSegments, 100000)
	return ca
}

func (ca *CpuAnalyzer) Start() error {
	errorLog := log.New(os.Stdout, "app", log.LstdFlags)
	var err error
	ca.esClient, err = elastic.NewClient(elastic.SetErrorLog(errorLog), elastic.SetURL(ca.cfg.GetEsHost()), elastic.SetSniff(false))
	if err != nil {
		ca.telemetry.Logger.Warn("new es client error", zap.Error(err))
	}
	info, code, err := ca.esClient.Ping(ca.cfg.GetEsHost()).Do(context.Background())
	if err != nil {
		ca.telemetry.Logger.Warn("new es client error", zap.Error(err))
	}
	ca.telemetry.Logger.Sugar().Infof("Es return with code %d and version %s", code, info.Version.Number)
	esversionCode, err := ca.esClient.ElasticsearchVersion(ca.cfg.GetEsHost())
	if err != nil {
		ca.telemetry.Logger.Warn("new es client error", zap.Error(err))
	}
	ca.telemetry.Logger.Sugar().Infof("es version %s\n", esversionCode)
	//go ca.SendTest()
	// go ca.SendCircle()
	go ca.ReceiveSendSignal()
	return nil
}

func (ca *CpuAnalyzer) ConsumeEvent(event *model.KindlingEvent) error {
	switch event.Name {
	case constnames.CpuEvent:
		ca.ConsumeCpuEvent(event)
	case constnames.JavaFutexInfo:
		ca.ConsumeJavaFutexEvent(event)
	case constnames.TransactionIdEvent:
		ca.ConsumeTransactionIdEvent(event)
	}
	return nil
}

func (ca *CpuAnalyzer) ConsumeTransactionIdEvent(event *model.KindlingEvent) {
	isEntry, _ := strconv.Atoi(event.GetStringUserAttribute("is_enter"))
	ev := &TransactionIdEvent{
		Timestamp: event.Timestamp,
		TraceId:   event.GetStringUserAttribute("trace_id"),
		IsEntry:   uint32(isEntry),
	}
	ca.PutEventToSegments(event.GetPid(), event.Ctx.ThreadInfo.GetTid(), event.Ctx.ThreadInfo.Comm, ev)
}

func (ca *CpuAnalyzer) ConsumeJavaFutexEvent(event *model.KindlingEvent) {
	ev := new(JavaFutexEvent)
	ev.StartTime = event.Timestamp
	for i := 0; i < int(event.ParamsNumber); i++ {
		userAttributes := event.UserAttributes[i]
		switch {
		case userAttributes.GetKey() == "end_time":
			ev.EndTime, _ = strconv.ParseUint(string(userAttributes.GetValue()), 10, 64)
		case userAttributes.GetKey() == "data":
			ev.DataVal = string(userAttributes.GetValue())
		}
	}
	ca.telemetry.Logger.Sugar().Infof("Receive a java futex event, %v", ev)
	ca.PutEventToSegments(event.GetPid(), event.Ctx.ThreadInfo.GetTid(), event.Ctx.ThreadInfo.Comm, ev)
	//ca.PutSegment(event.GetPid(), event.Ctx.ThreadInfo.GetTid(), event.Ctx.ThreadInfo.Comm, ev)
}

func (ca *CpuAnalyzer) ConsumeCpuEvent(event *model.KindlingEvent) {
	ev := new(CpuEvent)
	for i := 0; i < int(event.ParamsNumber); i++ {
		userAttributes := event.UserAttributes[i]
		switch {
		case userAttributes.GetKey() == "start_time":
			ev.StartTime = userAttributes.GetUintValue()
		case event.UserAttributes[i].GetKey() == "end_time":
			ev.EndTime = userAttributes.GetUintValue()
		case event.UserAttributes[i].GetKey() == "type_specs":
			ev.TypeSpecs = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "runq_latency":
			ev.RunqLatency = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "time_type":
			ev.TimeType = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "on_info":
			ev.OnInfo = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "off_info":
			ev.OffInfo = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "log":
			ev.Log = string(userAttributes.GetValue())
		case event.UserAttributes[i].GetKey() == "stack":
			ev.Stack = string(userAttributes.GetValue())
		}
	}
	ca.PutEventToSegments(event.GetPid(), event.Ctx.ThreadInfo.GetTid(), event.Ctx.ThreadInfo.Comm, ev)
	//ca.PutSegment(event.GetPid(), event.Ctx.ThreadInfo.GetTid(), event.Ctx.ThreadInfo.Comm, ev)
}

var nanoToSeconds uint64 = 1e9

func (ca *CpuAnalyzer) PutSegment(pid uint32, tid uint32, threadName string, event TimedEvent) {
	newTimeSegments := TimeSegments{
		Pid:      pid,
		Tid:      tid,
		BaseTime: event.StartTimestamp() / nanoToSeconds,
		Segments: NewCircleQueue(ca.cfg.GetSegmentSize() + 1),
	}
	segment := newSegment(pid, tid, threadName, newTimeSegments.BaseTime*nanoToSeconds, newTimeSegments.BaseTime*nanoToSeconds)
	segment.putTimedEvent(event)
	ca.SendSegment(*segment)
}

func (ca *CpuAnalyzer) PutEventToSegments(pid uint32, tid uint32, threadName string, event TimedEvent) {
	ca.lock.Lock()
	defer ca.lock.Unlock()
	tidCpuEvents, exist := ca.cpuPidEvents[pid]
	if !exist {
		if pid == 4777 {
			fmt.Println("reset tidCpuEvents:" + strconv.Itoa(int(pid)))
		}
		tidCpuEvents = make(map[uint32]TimeSegments)
		ca.cpuPidEvents[pid] = tidCpuEvents
	}
	timeSegments, exist := tidCpuEvents[tid]
	if exist {
		// The current segment is full
		if timeSegments.BaseTime+uint64(ca.cfg.GetSegmentSize()) <= event.StartTimestamp()/nanoToSeconds || event.EndTimestamp()/nanoToSeconds > timeSegments.BaseTime+uint64(ca.cfg.GetSegmentSize()) {
			clearSize := 0

			clearSize = ca.cfg.GetSegmentSize() / 2
			timeSegments.BaseTime = timeSegments.BaseTime + uint64(ca.cfg.GetSegmentSize())/2
			//if tid == 24917 {
			//	fmt.Println("-----------before clear----------")
			//	fmt.Println(time.Now().Unix())
			//	for i := 0; i < ca.cfg.GetSegmentSize() - 1; i++ {
			//		val, _ := timeSegments.Segments.GetByIndex(i)
			//		if val != nil {
			//			sg := val.(*Segment)
			//			fmt.Println("start:"+strconv.Itoa(int(sg.StartTime))+"   end:"+strconv.Itoa(int(sg.EndTime)))
			//		}
			//	}
			//	fmt.Println("----------------------------------")
			//}

			for i := 0; i < clearSize-1; i++ {
				val, _ := timeSegments.Segments.GetByIndex(i + ca.cfg.GetSegmentSize()/2)
				if val != nil {
					timeSegments.Segments.UpdateByIndex(i, val)
				}
				segmentTmp := newSegment(pid, tid, threadName,
					(timeSegments.BaseTime+uint64(i+clearSize))*nanoToSeconds,
					(timeSegments.BaseTime+uint64(i+clearSize+1))*nanoToSeconds)
				timeSegments.Segments.UpdateByIndex(i+clearSize, segmentTmp)
			}

			//if tid == 24917 {
			//	fmt.Println("-----------after clear----------")
			//	for i := 0; i < ca.cfg.GetSegmentSize() - 1; i++ {
			//		val, _ := timeSegments.Segments.GetByIndex(i)
			//		if val != nil {
			//			sg := val.(*Segment)
			//			fmt.Println("start:"+strconv.Itoa(int(sg.StartTime))+"   end:"+strconv.Itoa(int(sg.EndTime)))
			//		}
			//	}
			//	fmt.Println("----------------------------------")
			//}

		}
		if int(event.EndTimestamp()/nanoToSeconds-timeSegments.BaseTime) < 0 {
			return
		}
		// 开始时间基于baseTime的偏移量
		startOffset := int(event.StartTimestamp()/nanoToSeconds - timeSegments.BaseTime)
		// 结束时间基于开始时间的偏移量
		endOffset := int(event.EndTimestamp()/nanoToSeconds - event.StartTimestamp()/nanoToSeconds)
		if startOffset < 0 {
			startOffset = 0
			endOffset = int(event.EndTimestamp()/nanoToSeconds - timeSegments.BaseTime)
			if endOffset < 0 {
				fmt.Println("start<end")
			}
		}
		if tid == 4777 {
			fmt.Println("-----------put----------")
			fmt.Println("start:" + strconv.Itoa(int(event.StartTimestamp())) + "   end:" + strconv.Itoa(int(event.EndTimestamp())))
			fmt.Println("----------------------------------")
		}
		for i := startOffset; i <= startOffset+endOffset; i++ {
			var segment *Segment
			val, _ := timeSegments.Segments.GetByIndex(int(i))
			if val == nil {
				segment = newSegment(pid, tid, threadName,
					(timeSegments.BaseTime+uint64(i))*nanoToSeconds,
					(timeSegments.BaseTime+uint64(i+1))*nanoToSeconds)
			} else {
				segment = val.(*Segment)
			}
			segment.putTimedEvent(event)
			segment.IsSend = 0
			if tid == 4777 {
				fmt.Println("basetime:" + strconv.Itoa(int(timeSegments.BaseTime)) + "    i:" + strconv.Itoa(int(i)))
			}
			timeSegments.Segments.UpdateByIndex(int(i), segment)
			tidCpuEvents[tid] = timeSegments
		}

	} else {
		newTimeSegments := TimeSegments{
			Pid:      pid,
			Tid:      tid,
			BaseTime: event.StartTimestamp() / nanoToSeconds,
			Segments: NewCircleQueue(ca.cfg.GetSegmentSize() + 1),
		}
		for i := 0; i < ca.cfg.GetSegmentSize(); i++ {
			segment := newSegment(pid, tid, threadName,
				(timeSegments.BaseTime+uint64(i))*nanoToSeconds,
				(timeSegments.BaseTime+uint64(i+1))*nanoToSeconds)
			newTimeSegments.Segments.Push(segment)
		}
		val, _ := newTimeSegments.Segments.GetByIndex(0)
		segment := val.(*Segment)
		segment.putTimedEvent(event)
		tidCpuEvents[tid] = newTimeSegments
	}
}

func (ca *CpuAnalyzer) Shutdown() error {
	// TODO: implement
	return nil
}

File: refactor/camera-agent-main/src/cpu_analyzer/cpu_analyzer.rs
use crate::camera_exporter::camera_exporter::CameraExporter;
use crate::cpu_analyzer::circle_queue::CircleQueue;
use crate::cpu_analyzer::model::TimedEventKind::TimedCpuEventKind;
use crate::cpu_analyzer::model::{CpuEvent, JavaFutexEvent, TimeSegments, TransactionIdEvent};
use crate::cpu_analyzer::model::{Segment, TimedEvent};
use crate::cpu_analyzer::time_aggregate::{get_metric_name, CpuTimeType};
use crate::probe_to_rust::KindlingEventForGo;
use byteorder::{LittleEndian, ReadBytesExt};
use log::{debug, info};
use std::collections::HashMap;
use std::io::Cursor;
use std::sync::{Arc, Mutex};
use std::time::{Duration, SystemTime};

use super::tid_exit_clear::DeleteTid;
use super::time_aggregate::AggregatedTime;
use crate::metric_exporter::{self};
use crate::model::consts;
use crate::traceid_analyzer::{SignalEvent, TraceEvent};
use opentelemetry_api::KeyValue;
const NANO_TO_SECONDS: u64 = 1_000_000_000;
const CHECK_INTERVAL: Duration = Duration::from_secs(1);

pub(crate) const MAX_SEGMENT_SIZE: usize = 40;

pub(crate) const UPPER_LIMITE_TS: u64 = 2000000000000000000;

pub(crate) const LOWER_LIMITE_TS: u64 = 1600000000000000000;

pub struct CpuAnalyzer {
    pub cpu_pid_events: HashMap<u32, HashMap<u32, TimeSegments>>,
    pub tid_exit_queue: Vec<DeleteTid>,
    pub metric_signals: Vec<SignalEvent>,
    pub metric_exporter: Arc<dyn metric_exporter::MetricExporter>,
    pub profile_exporter: Arc<CameraExporter>,
}

pub fn delay_send_metrics(cca: &Arc<Mutex<CpuAnalyzer>>) {
    let mut cpu_analyzer = cca.lock().unwrap();
    cpu_analyzer.send_remove_metrics();
}

pub fn clear_tid(cca: &Arc<Mutex<CpuAnalyzer>>, expired_duration: Duration) {
    let mut ca_guard = cca.lock().unwrap();
    ca_guard.tid_delete(expired_duration);
}

pub fn consume_cpu_event(event: &KindlingEventForGo, cca: &Arc<Mutex<CpuAnalyzer>>) {
    let mut ev = CpuEvent::default();
    for i in 0..event.params_number as usize {
        let user_attributes = event.user_attributes[i];
        match i {
            0 => ev.start_time = user_attributes.get_uint_value(),
            1 => ev.end_time = user_attributes.get_uint_value(),
            2 => {
                let val = user_attributes.get_value().unwrap();
                ev.type_specs = read_u64_values(val);
            }
            3 => {
                let val = user_attributes.get_value().unwrap();
                ev.runq_latency = read_u64_values(val);
            }
            4 => {
                let val = user_attributes.get_value().unwrap();
                ev.time_type = read_u8_values(val);
            }
            5 => ev.stack = read_string_value(user_attributes.get_value()),
            6 => ev.log = read_string_value(user_attributes.get_value()),
            7 => ev.on_info = read_string_value(user_attributes.get_value()),
            8 => ev.off_info = read_string_value(user_attributes.get_value()),
            _ => (),
        }
    }

    if ev.start_time < LOWER_LIMITE_TS {
        return;
    }

    // if event.get_pid() != 9689 as u32  && event.get_pid() != 4509 as u32 && event.get_pid() != 26600 as u32{
    //     info!("{}", &ev);
    //     info!("tid: {}, thread name: {}", event.get_tid(),&event.get_comm())
    // }

    let mut ca_guard = cca.lock().unwrap();
    if let Some((start, end, segments)) =
        ca_guard.put_event_to_segments(event.get_pid(), event.get_tid(), &event.get_comm(), &ev)
    {
        let mut corrected_start = start;
        if corrected_start < 0 {
            corrected_start = 0;
        }
        for i in corrected_start..end {
            if let Some(segment) = segments.get_by_index_mut(i as usize) {
                segment.put_cpu_event(ev.clone());
                segment.is_send = 0;
            }
        }
        if let Some(segment) = segments.get_by_index_mut(end as usize) {
            segment.put_cpu_event(ev);
            segment.is_send = 0;
        }
    }
}

pub fn consume_java_futex_event(event: &KindlingEventForGo, cca: &Arc<Mutex<CpuAnalyzer>>) {
    let ev = JavaFutexEvent {
        start_time: event.timestamp,
        end_time: event.user_attributes[0].get_uint_value(),
        data_val: read_string_value(event.user_attributes[1].get_value()),
    };
    //debug!("{}", &ev);

    let mut ca_guard = cca.lock().unwrap();
    if let Some((start, end, segments)) =
        ca_guard.put_event_to_segments(event.get_pid(), event.get_tid(), &event.get_comm(), &ev)
    {
        for i in start..end {
            if let Some(segment) = segments.get_by_index_mut(i as usize) {
                segment.put_java_futex_event(ev.clone());
                segment.is_send = 0;
            }
        }
        if let Some(segment) = segments.get_by_index_mut(end as usize) {
            segment.put_java_futex_event(ev);
            segment.is_send = 0;
        }
    }
}

pub fn consume_transaction_id_event(event: &KindlingEventForGo, cca: &Arc<Mutex<CpuAnalyzer>>) {
    let ev = TransactionIdEvent {
        timestamp: event.timestamp,
        trace_id: read_string_value(event.user_attributes[0].get_value()),
        is_entry: read_string_value(event.user_attributes[1].get_value()),
    };
    info!("{}", &ev);

    let mut ca_guard = cca.lock().unwrap();
    if let Some((start, end, segments)) =
        ca_guard.put_event_to_segments(event.get_pid(), event.get_tid(), &event.get_comm(), &ev)
    {
        for i in start..end {
            if let Some(segment) = segments.get_by_index_mut(i as usize) {
                segment.put_transaction_id_event(ev.clone());
                segment.is_send = 0;
            }
        }
        if let Some(segment) = segments.get_by_index_mut(end as usize) {
            segment.put_transaction_id_event(ev);
            segment.is_send = 0;
        }
    }
}

pub fn consume_procexit_event(event: &KindlingEventForGo, ca: &Arc<Mutex<CpuAnalyzer>>) {
    let pid = event.get_pid();
    let tid = event.get_tid();
    debug!("Receive a procexit event: pid={}, tid={}", pid, tid);
    let mut ca_guard = ca.lock().unwrap();
    ca_guard.trim_exited_thread(pid, tid);
}

impl CpuAnalyzer {
    pub fn new(
        metric_exporter: Arc<dyn metric_exporter::MetricExporter>,
        profile_exporter: Arc<CameraExporter>,
    ) -> Self {
        CpuAnalyzer {
            cpu_pid_events: HashMap::new(),
            tid_exit_queue: Vec::new(),
            metric_signals: Vec::new(),
            metric_exporter,
            profile_exporter,
        }
    }

    pub fn handle_profiling_signal(&mut self, event: &TraceEvent) {
        if event.is_normal {
            self.send_event(
                event.pid,
                event.tid,
                event.end_time - event.duration,
                event.end_time,
                event,
            );
        } else {
            self.send_events(
                event.pid,
                event.end_time - event.duration,
                event.end_time,
                event,
            );
        }
    }

    pub fn handle_metric_signal(&mut self, mut events: Vec<SignalEvent>) {
        events.drain(..).for_each(|event| {
            debug!("signal metric: {:?}", &event);
            if self.send_metric(&event).is_none() {
                // Cache SignalEvent to DelayQueue.
                self.cache_metric_signal(event);
            }
        });
    }

    pub fn send_remove_metrics(&mut self) {
        if self.metric_signals.is_empty() {
            return;
        }

        let mut removed_indices = vec![];
        // Send Metrics
        for (i, value) in self.metric_signals.iter().enumerate() {
            if self.send_metric(value).is_some() {
                removed_indices.push(i);
            }
        }

        let mut offset = 0;
        // Remove Sent Metrics
        self.metric_signals.retain(|_| {
            let current_index = offset;
            offset += 1;
            !removed_indices.contains(&current_index)
        });
    }

    fn cache_metric_signal(&mut self, event: SignalEvent) {
        self.metric_signals.push(event);
    }

    fn send_metric(&self, event: &SignalEvent) -> Option<AggregatedTime> {
        let agg = self.get_times(event.pid, event.tid, event.start_time, event.end_time)?;
        debug!("Metric Aggregate: {:?}", agg);
        let attributes = vec![
            KeyValue::new(consts::PID, event.pid.to_string()),
            KeyValue::new(consts::CONTENT_KEY, event.content_key.clone()),
            KeyValue::new(consts::CONTAINER_ID, event.container_id.clone()),
        ];
        for i in 0..agg.len() {
            let time_type = CpuTimeType::from_usize(i).unwrap();
            let duration = agg.times[i];
            let duration_name = get_metric_name(&time_type, consts::PROFILING_DURATION_SUFFIX);
            self.metric_exporter
                .record_metric(duration_name, duration, &attributes);
            let count = agg.counts[i];
            let count_name = get_metric_name(&time_type, consts::PROFILING_COUNT_SUFFIX);
            self.metric_exporter
                .record_metric(count_name, count, &attributes);
        }
        if agg.runq_duration > 0 {
            self.metric_exporter.record_metric(
                consts::PROFILING_RUNQ_DURATION_METRIC_NAME,
                agg.runq_duration,
                &attributes,
            );
        }
        Some(agg)
    }

    pub fn put_event_to_segments(
        &mut self,
        pid: u32,
        tid: u32,
        thread_name: &str,
        event: &dyn TimedEvent,
    ) -> Option<(i32, i32, &mut CircleQueue)> {
        if event.end_timestamp() > UPPER_LIMITE_TS || event.start_timestamp() > UPPER_LIMITE_TS {
            info!(
                "filiter err time event!  start: {:?}   end: {:?}    kind: {:?}!",
                event.start_timestamp(),
                event.end_timestamp(),
                event.kind()
            );
            return None;
        }
        let tid_cpu_events = self.cpu_pid_events.entry(pid).or_insert_with(HashMap::new);
        let time_segments = tid_cpu_events.entry(tid).or_insert_with(|| {
            let base_time = event.start_timestamp() / NANO_TO_SECONDS;
            let segments = create_initial_segments(base_time);
            TimeSegments::new(pid, tid, thread_name.to_string(), base_time, segments)
        });
        if event.kind() == TimedCpuEventKind {
            // Set LatestTime.
            time_segments.update_time(event.end_timestamp());
        }

        if event.end_timestamp() / NANO_TO_SECONDS < time_segments.base_time {
            return None;
        }
        let mut end_offset =
            (event.end_timestamp() / NANO_TO_SECONDS - time_segments.base_time) as i32;

        let mut start_offset =
            (event.start_timestamp() / NANO_TO_SECONDS - time_segments.base_time) as i32;
        let should_clear_segments =
            start_offset >= MAX_SEGMENT_SIZE as i32 || end_offset > MAX_SEGMENT_SIZE as i32;

        if should_clear_segments {
            if start_offset * 2 >= 3 * MAX_SEGMENT_SIZE as i32 {
                time_segments.segments.clear();
                time_segments.base_time = event.start_timestamp() / NANO_TO_SECONDS;
                end_offset -= start_offset;
                start_offset = 0;
                time_segments.segments = create_initial_segments(time_segments.base_time);
            } else if end_offset > MAX_SEGMENT_SIZE as i32 {
                time_segments.segments.clear();
                time_segments.base_time =
                    (event.end_timestamp() / NANO_TO_SECONDS) - ((MAX_SEGMENT_SIZE / 2) as u64);
                start_offset = ((event.start_timestamp() / NANO_TO_SECONDS)
                    - time_segments.base_time as u64) as i32;
                end_offset = (MAX_SEGMENT_SIZE / 2) as i32;
                time_segments.segments = create_initial_segments(time_segments.base_time);
            } else {
                let clear_size = MAX_SEGMENT_SIZE / 2;
                time_segments.base_time += clear_size as u64;
                if start_offset < clear_size as i32 {
                    start_offset = 0;
                } else {
                    start_offset -= clear_size as i32;
                }
                end_offset -= clear_size as i32;
                for i in 0..clear_size {
                    let moved_index = i + clear_size;
                    if let Some(_segment) = time_segments.segments.get_by_index(moved_index) {
                        time_segments.segments.update_by_moved_index(i, moved_index);
                    }
                    let segment_tmp = Segment::new(
                        (time_segments.base_time + (moved_index as u64)) * NANO_TO_SECONDS,
                        (time_segments.base_time + ((moved_index + 1) as u64)) * NANO_TO_SECONDS,
                    );
                    time_segments
                        .segments
                        .update_by_index(moved_index, segment_tmp);
                }
            }
        }
        time_segments.update_thread_name(thread_name);

        end_offset = end_offset.min(MAX_SEGMENT_SIZE as i32 - 1);
        Some((start_offset, end_offset, &mut time_segments.segments))
    }

    pub fn send_event(
        &mut self,
        pid: u32,
        tid: u32,
        start_time: u64,
        end_time: u64,
        trace_event: &TraceEvent,
    ) {
        let tid_cpu_events = match self.cpu_pid_events.get_mut(&pid) {
            Some(tid_cpu_events) => tid_cpu_events,
            None => {
                return;
            }
        };
        let time_segments = match tid_cpu_events.get_mut(&tid) {
            Some(time_segments) => time_segments,
            None => {
                return;
            }
        };

        let start_time_second = start_time / NANO_TO_SECONDS;
        let end_time_second = end_time / NANO_TO_SECONDS;

        if end_time_second < time_segments.base_time
            || start_time_second > time_segments.base_time + (MAX_SEGMENT_SIZE as u64)
        {
            return;
        }

        let start_index = (start_time_second - time_segments.base_time) as i32;
        let end_index =
            (end_time_second - time_segments.base_time).min(MAX_SEGMENT_SIZE as u64) as i32;

        let thread_name = time_segments.thread_name.as_str();
        for i in start_index..=end_index {
            if let Some(segment) = time_segments.segments.get_by_index_mut(i as usize) {
                if segment.is_not_empty() {
                    segment.update_index_timestamp();
                    // Send To Es.
                    let profiling_event = segment.to_data_group(pid, tid, thread_name, trace_event);
                    self.profile_exporter.consume_profiling(profiling_event);
                    debug!("{:?}", segment);
                    segment.is_send = 1;
                }
            }
        }
    }

    pub fn send_events(
        &mut self,
        pid: u32,
        start_time: u64,
        end_time: u64,
        trace_event: &TraceEvent,
    ) {
        let tid_cpu_events = match self.cpu_pid_events.get_mut(&pid) {
            Some(tid_cpu_events) => tid_cpu_events,
            None => {
                return;
            }
        };

        let start_time_second = start_time / NANO_TO_SECONDS;
        let end_time_second = end_time / NANO_TO_SECONDS;

        for time_segments in tid_cpu_events.values_mut() {
            if end_time_second < time_segments.base_time
                || start_time_second > time_segments.base_time + (MAX_SEGMENT_SIZE as u64)
            {
                continue;
            }

            let start_index = (start_time_second - time_segments.base_time) as i32;
            let end_index =
                (end_time_second - time_segments.base_time).min(MAX_SEGMENT_SIZE as u64) as i32;

            let pid = time_segments.pid;
            let tid = time_segments.tid;
            let thread_name = time_segments.thread_name.as_str();
            for i in start_index..=end_index {
                if let Some(segment) = time_segments.segments.get_by_index_mut(i as usize) {
                    if segment.is_not_empty() {
                        segment.update_index_timestamp();
                        // Send To Es.
                        let profiling_event =
                            segment.to_data_group(pid, tid, thread_name, trace_event);
                        self.profile_exporter.consume_profiling(profiling_event);
                        debug!("{:?}", segment);
                        segment.is_send = 1;
                    }
                }
            }
        }
    }

    pub fn trim_exited_thread(&mut self, pid: u32, tid: u32) {
        let cache_elem = DeleteTid {
            pid,
            tid,
            exit_time: SystemTime::now(),
        };

        self.tid_exit_queue.push(cache_elem);
    }

    pub fn tid_delete(&mut self, expired_duration: Duration) {
        let now = SystemTime::now();

        loop {
            if let Some(elem) = self.tid_exit_queue.first() {
                if elem.exit_time + expired_duration > now {
                    break;
                }

                if let Some(tid_events_map) = self.cpu_pid_events.get_mut(&elem.pid) {
                    tid_events_map.remove(&elem.tid);
                }

                self.tid_exit_queue.remove(0);
            } else {
                break;
            }
        }
    }
}

fn read_u64_values(val: &[u8]) -> Vec<u64> {
    let mut cursor = Cursor::new(val);
    let count = val.len() / 8;
    let mut values = Vec::with_capacity(count);
    for _ in 0..count {
        values.push(cursor.read_u64::<LittleEndian>().unwrap());
    }
    values
}

fn read_u8_values(val: &[u8]) -> Vec<u8> {
    val.to_vec()
}

fn read_string_value(val: Option<&[u8]>) -> String {
    match val {
        Some(bytes) => {
            let mut string = String::with_capacity(bytes.len());
            for &byte in bytes {
                string.push(byte as char);
            }

            string
        }
        None => String::new(),
    }
}

fn create_initial_segments(base_time: u64) -> CircleQueue {
    let mut segments = CircleQueue::new(MAX_SEGMENT_SIZE);
    for i in 0..MAX_SEGMENT_SIZE {
        let segment = Segment::new(
            (base_time + (i as u64)) * NANO_TO_SECONDS,
            (base_time + (i as u64) + 1) * NANO_TO_SECONDS,
        );
        segments.update_by_index(i, segment);
    }
    segments
}
